### ISSUE TITLE
strictly require level to be passed in, drop level deps

### ISSUE THREAD

**Issue by mikeal** [ethereumjs/merkle-patricia-tree#108]
I’m seeing some nastly browser bundling issues because this package is in my dep tree and depends on modules which have ancient versions of level and there have been several breaking changes since.

These level deps are pretty massive and unnecessary deps since they only get used when a level instance is not passed in. If you moved to require this explicitly you could drop the deps.

In the meantime I’m going to have to drop ethereum support from my default set so that I can avoid this issue.

**Comment by holgerd77** [comment]
@ryanio What do you think about this? Would it be a way to move the `level` dependencies to the `devDependencies` and then just do the testing with the libs and otherwise require people to pass its own instance? Not seeing the downside atm apart from some (minor) convenience loss.

We should then also state in the docs what level version is tested to be working with the library. Eventually we can also do some update on this occasion.

What do you think (if you generally think this being a good idea) should we integrate into the followup major release after having merged #107 ?

@mikeal Thanks for reporting! Some context, we are relatively close to release the TypeScript + Promisified version of the library. Any suggestions on this welcome as well!

**Comment by mikeal** [comment]
> Some context, we are relatively close to release the TypeScript + Promisified version of the library. Any suggestions on this welcome as well!

We moved all of IPLD and IPFS to use async/await, so we’d be happy to take a promise version of the library ;)

**Comment by ryanio** [comment]
Hey, sounds like a good idea :) I'll put it on my todo list for when after #107 is merged.

To add to the discussion, the following is from a set of recent performance [recommendations](https://github.com/EthWorks/ethereumjs-vm-benchmarks/blob/master/results/results.md#remove-level-make-merkle-patricia-tree-a-sync-structure) from EthWorks for `ethereumjs-vm`:

> #### Remove level, make merkle-patricia-tree a sync structure
>
> Unfortunately performance impact is hard to measure because of the current async nature. Since merkle-patricia-tree is such a core component a large performance benefit is expected. At minimum it would eliminate 1.7% of time spent on constructing unused errors.
>
>Level should be replaced with a simple map. The only operation that could potentially be slower is committing checkpoints, which would be replaced by a simple for (copy each of the values from ScratchDB to DB). Note that for the keys to work they might need to be stringified, but this should be less costly than encoding and decoding the values each time they are read/saved from the database.
>
>A ton of overhead would be removed from the structure - since map operations are synchronous there would be no need for callbacks and semaphores. This in turn would massively simplify the code of the data structure.
>
>Level is responsible for one of the biggest performance drain when creating new contracts which is it's errno dependency. Calling db.get (levelup@1.3.9 /lib/levelup.js:195) in levelup sometimes results in errors. The program spends 1.7% of the total runtime just constructing those errors.
>
>Probably caused by weird stack trace manipulations and a ton of information on the error object. I suspect that this information isn't used in any meaningful way.
>
>The total count of errors created during the 60s benchmark was: 46529. Constructing 46k errors should take ~150ms, not upwards of a second.
>
> Overall benefits
> * cleaner, easier to reason about code
> * no async - faster code
> * no semaphores
> * no errno error shenanigans

I'm not sure how much of the above changes due to #107, but can still investigate the core recommendation that "level should be replaced with a simple map." @s1na do you have any thoughts to add?

**Comment by mikeal** [comment]
> make merkle-patricia-tree a sync structure

I can see how you could move the data structure to an inmemory sync structure but you’re still going to end up with async methods for creating it since the browser’s hashing functions are async.

We’ve done similar audits that found “unnecessary async” to be a perf problem and have factored a lot of it out, which is why things like block serialization/deserialization are now synchronous, but we still have to stick async API’s everywhere we’re doing a hash in order to keep working in the browser.

If you can avoid shipping `level` stuff you should, it’ll definitely be faster. `js-ipfs` just dropped level on the browser side to use IndexedDB directly and that was something like a 40% perf gain, and a smaller bundle size. But in Node.js it’s still the best option for a flexible on-disc embedded database interface.

P.S. we’ve also been very happy with factoring out all the streams code for async generators, in case that comes up for you elsewhere 😉

**Comment by holgerd77** [comment]
@mikeal thanks, that's really valuable input! 😊

**Comment by sz-piotr** [comment]
>  can see how you could move the data structure to an inmemory sync structure but you’re still going to end up with async methods for creating it since the browser’s hashing functions are async.

What hashing functions are async in the browser?

**Comment by mikeal** [comment]
SHA2 ;) https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

**Comment by kumavis** [comment]
@mikeal yeah raw keccak is unlikely to get browser support, so less relevant here
and I haven't found a keccak package with async native bindings, but might exist, could give us some parallelism gains if we adopt lazy hashing

I've advocated using ipfs's [`interface-datastore`](https://github.com/ipfs/interface-datastore/) in place of level, do you think its appropriate?

I think dropping level as reccomended here is a nice minimal fix :+1:
of course needs to be behind a proper major version bump

**Comment by mikeal** [comment]
> raw keccak is unlikely to get browser support

Sure, but we have a multihash interface that supports several hashing functions. Once one of them is async the interface has to be async ;)

**Comment by ryanio** [comment]
The only two level deps `level-mem` and `level-ws` are in use and updated to latest in 4.0.0 release (#111).

@mikeal could you give some feedback in #117 if providing a browser build would be helpful for you?
